src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs

```diff
@@ -97,74 +145,163 @@ public static string TryCompileSectionNameToRegEx(string sectionName)
                         throw ExceptionUtilities.UnexpectedValue(tokenKind);
                 }
             }
-            sb.Append('$');
+            // If we're parsing a choice we should not exit without a closing '}'
+            return !parsingChoice;
+        }
 
-            return sb.ToString();
+        private static bool TryCompileChoice(ref SectionNameLexer lexer, StringBuilder sb)
+        {
+            // Start a non-capturing group for the choice
+            sb.Append("(?:");
+
+            // We start immediately after a '{'
+            // Try to compile the nested <path-list>
+            while (TryCompilePathList(ref lexer, sb, parsingChoice: true))
+            {
+                // If we've succesfully compiled a <path-list> the last token should
+                // have been a ',' or a '}'
+                char lastChar = lexer[lexer.Position - 1];
+                if (lastChar == ',')
+                {
+                    // Another option
+                    sb.Append("|");
+                }
+                else if (lastChar == '}')
+                {
+                    // Close out the capture group
+                    sb.Append(")");
+                    return true;
+                }
+                else
+                {
+                    throw ExceptionUtilities.UnexpectedValue(lastChar);
+                }
+            }
+
+            // Propagate failure
+            return false;
+        }
+
+        private static (string numStart, string numEnd)? TryParseNumberRange(ref SectionNameLexer lexer)
+        {
+            var saved = lexer.Position;
+            // We start immediately after a '{'
+            var num1 = lexer.TryLexNumber();
+            if (num1 is null || lexer.IsDone)
+            {
+                // Not a number
+                lexer.Position = saved;
+                return null;
+            }
+
+            // The next two characters must be ".."
+            if (lexer.EatCurrentCharacter() != '.' || lexer.IsDone ||
+                lexer.EatCurrentCharacter() != '.' || lexer.IsDone)
+            {
+                lexer.Position = saved;
+                return null;
+            }
+
+            // Now another number
+            var num2 = lexer.TryLexNumber();
+            if (num2 is null || lexer.IsDone ||
+                lexer.Lex() != TokenKind.CloseCurly)
```

Comment:
Seems a bit inconsistent to validate the close curly here but not the open one. Would expect this parsing routine to be completely ignorant of curlies or verify both.